src/core/thread/mesh_forwarder.cpp

```diff
@@ -1287,7 +1397,15 @@ void MeshForwarder::HandleFragment(uint8_t *               aFrame,
         if (message->GetOffset() >= message->GetLength())
         {
             mReassemblyList.Dequeue(*message);
-            HandleDatagram(*message, aLinkInfo, aMacSource);
+
+            if (aReceive)
+            {
+                HandleDatagram(*message, aLinkInfo, aMacSource);
+            }
+            else
+            {
+                message->Free();
```

Comment:
Routers simply forwarding 6LoWPAN fragments are now required to reassemble the IPv6 datagram.  However, all that's really needed is the priority for a given datagram tag.  Can we update this so we are only storing at most one message buffer for each datagram tag?

Reply:
In current code, when calling SetLength(), the message will malloc buffers. I may need to modify the logic of this function to resolve this issue.

Reply:
I am not in favor of changing `Message::SetLength()` behavior. `Message` is well-defined and well-tested class. Other `Message` class methods like `Read()` and `Write()` allow access to buffer up to its previously set length. With this change users of `Message` should know and treat  the APIs differently.